### ISSUE TITLE
Connection pool timeout errors when using higher connection_limit

### ISSUE THREAD

**Issue by gnetty** [prisma/prisma#9622]
### Bug description

We’ve been experiencing intermittent “Timed out fetching a new connection from the pool" errors. Our connection_limit is set to 65 and pool_timeout was 6.

This is what we’ve observed:
- We see many of these errors over the course of a few minutes
- The CPU and memory usage of the application during the time are both under 50%
- The errors stop a few minutes later and database calls start working again
- At the time we get these errors, our Postgres database shows fewer connections than the connection_limit in Prisma

We have tried adjusting the pool_timeout and connection_limit, with some success. Strangely, reducing the connection_limit seems to help.

### How to reproduce

These errors seem to occur during periods of higher traffic in our app.

### Expected behavior

I expected a higher connection_limit to reduce the number of these exceptions. I also expected the database to report the number of active connections equal to our connection_limit size.

### Prisma information

not query specific

### Environment & setup

- OS: Alpine Linux (Docker container in Azure App Service)
- Database: PostgreSQL (Azure)
- Node.js version: 12.14.1

### Prisma Version

```
prisma               : 2.22.1
@prisma/client       : 2.22.1
Current platform     : darwin
Query Engine         : query-engine 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash : 60cc71d884972ab4e897f0277c4b84383dddaf6c
Studio               : 0.379.0
Preview Features     : orderByRelation
```

**Comment by shortcircuit3** [comment]
We're also experiencing this issue. I'm forced to restart our production app every 30 minutes to reset connections :(

**Comment by garrensmith** [comment]
@shortcircuit3 we are looking at this issue at the moment.

**Comment by gnetty** [comment]
@shortcircuit3, we are now suspecting that our problem is actually due to intermittent network issues between our app and the database.

We've replicated network outages and it seems that the connection pool timeout error above could be misleading. We noticed that if the pool is already warm (with active connections) and we disable the network, then we receive the error above (timed out fetching a connection from the pool). However, if the pool is empty or a new connection is attempted, then we correctly receive a "can't reach database server" error. 

Bottom line is that the connection pool timeout error might not tell the full story.

**Comment by janpio** [comment]
(You can probably get around that behavior by disabling the timeout completely @gnetty - `pool_timeout=0` should do the trick in the connection string. This might of course have other, undesirable effects on your application behavior.)

**Comment by garrensmith** [comment]
This is fixed. The full information can be found here https://github.com/importcjj/mobc/issues/63
This will be out in the next release. You can test it now with prisma@dev

**Comment by flazouh** [comment]
Hi, I just got:
`
Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 1
`
and was wondering if you help me with that, i'm not a devos engineer, just someone that has a sass and is using prisma+supabase for his database.
Here:

```js
datasource db {
  provider  = "postgresql"
  url       = env("DATABASE_URL")
  directUrl = env("DIRECT_URL")
}
```
My DATABASE_URL is 
`
postgres://postgres.<appname>:<password>@aws-0-eu-west-2.pooler.supabase.com:6543/postgres?pgbouncer=true
`

**Comment by bakigultirport** [comment]
> Hi, I just got: `Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 1` googled it, and found this github issue and was wondering if you help me with that, i'm not a devos engineer, just someone that has a sass and is using prisma+supabase for his database. Here:
> 
> ```js
> datasource db {
>   provider  = "postgresql"
>   url       = env("DATABASE_URL")
>   directUrl = env("DIRECT_URL")
> }
> ```
> 
> My DATABASE_URL is `postgres://postgres.<appname>:<password>@aws-0-eu-west-2.pooler.supabase.com:6543/postgres?pgbouncer=true`
> 
> I get this everyday and I have to restart my prod supabase database, what I tried to do is the following code:
> 
> ```ts
> import { Injectable, NestMiddleware, Logger } from '@nestjs/common';
> import { Request, Response, NextFunction } from 'express';
> import { Prisma } from '@prisma/client';
> import { PrismaService } from '../prisma/prisma.service'; // Make sure this path is correct
> import { AxiomEvent } from 'src/usage-guard/AxiomEvent';
> 
> @Injectable()
> export class PrismaErrorMiddleware implements NestMiddleware {
> 	private readonly logger = new Logger(PrismaErrorMiddleware.name);
> 
> 	constructor(private prismaService: PrismaService) {}
> 
> 	async use(req: Request, res: Response, next: NextFunction) {
> 		try {
> 			await next();
> 		} catch (err: unknown) {
> 			this.logger.error(`Error in PrismaErrorMiddleware: ${err instanceof Error ? err.stack : err}`);
> 
> 			if (
> 				err instanceof Prisma.PrismaClientInitializationError ||
> 				(err instanceof Prisma.PrismaClientKnownRequestError && err.code === 'P2024')
> 			) {
> 				this.logger.error(`${AxiomEvent.RESETTING_PRISMA_CONNECTION}: Prisma connection error: ${err.message}`);
> 				
> 				// Disconnect Prisma
> 				await this.prismaService.$disconnect();
> 				
> 				// Attempt to reconnect
> 				try {
> 					await this.prismaService.$connect();
> 					this.logger.log('Successfully reconnected to Prisma');
> 					
> 					// Retry the original request
> 					await this.retryRequest(req, res, next);
> 				} catch (reconnectError) {
> 					this.logger.error(`Failed to reconnect to Prisma: ${reconnectError}`);
> 					res.status(503).json({
> 						statusCode: 503,
> 						message: 'Database connection error. Please try again later.',
> 					});
> 				}
> 			} else {
> 				// Pass other errors to the global error handler
> 				next(err);
> 			}
> 		}
> 	}
> 
> 	private async retryRequest(req: Request, res: Response, next: NextFunction): Promise<void> {
> 		return new Promise((resolve, reject) => {
> 			this.logger.log('Retrying the original request');
> 			const retryNext = (err?: any) => {
> 				if (err) {
> 					reject(err);
> 				} else {
> 					resolve();
> 				}
> 			};
> 			next(retryNext);
> 		});
> 	}
> }
> ```
> 
> Which is a solution I found with the help of claude sonnet 3.5, the code makes sense, but I have yet to try it in prod.

Hey, This solution seems interesting, did it  work for you.
